File: bindings-js/core/src/viewport.ts

// TODO: auto-generate these types from the Rust side: https://github.com/keller-mark/pluot/issues/133
export type AspectRatioMode = "Ignore" | "Contain" | "Cover";
export type AspectRatioAlignmentMode = "Center" | "Start" | "End";

export type Margins = {
  marginTop?: number;
  marginRight?: number;
  marginBottom?: number;
  marginLeft?: number;
};

export type ViewportParams = {
  width: number;
  height: number;
  aspectRatioMode: AspectRatioMode;
  aspectRatioAlignmentMode: AspectRatioAlignmentMode;
  margins?: Margins;
};

export type Bounds = {
  // Each value is optional.
  // When an entire dimension is omitted (X or Y),
  // use the current camera settings for that dimension.
  // When a single value is omitted (e.g., xMin), ensure the resulting camera matrix keeps this boundary unchanged.
  xMin?: number;
  xMax?: number;
  yMin?: number;
  yMax?: number;
};

/**
 * Calculate the visible data range based on camera view and viewport parameters.
 *
 * Bounds are expressed in normalized data coordinates where 0.0 is the
 * left/top edge of the data and 1.0 is the right/bottom edge.
 *
 * @param cameraMatrix - The current camera matrix, typically from the renderer's
 *   current state.
 * @param viewportParams - Describes the canvas size, aspect-ratio handling, and
 *   margins.
 * @returns The visible data range as `{ xMin, xMax, yMin, yMax }` in normalized
 *   data coordinates.
 *
 * @example
 * ```ts
 * import { getBounds } from "@pluot/core";
 *
 * const viewport = {
 *   width: 800,
 *   height: 600,
 *   aspectRatioMode: "Contain",
 *   aspectRatioAlignmentMode: "Center",
 *   margins: { marginTop: 10, marginRight: 10, marginBottom: 10, marginLeft: 10 },
 * };
 *
 * // Identity camera matrix (full data range visible).
 * const identity = new Float32Array([
 *   1, 0, 0, 0,
 *   0, 1, 0, 0,
 *   0, 0, 1, 0,
 *   0, 0, 0, 1,
 * ]);
 *
 * const bounds = getBounds(identity, viewport);
 * // bounds.xMin, bounds.xMax, bounds.yMin, bounds.yMax are all in [0, 1]
 * // for the identity matrix with "Contain" mode.
 * ```
 */
export function getBounds(cameraMatrix: Float32Array, viewportParams: ViewportParams): Required<Bounds> {
  const zoomX = cameraMatrix[0];
  const zoomY = cameraMatrix[5];
  const translateX = cameraMatrix[12];
  const translateY = cameraMatrix[13];

  const marginTop = viewportParams.margins?.marginTop ?? 0;
  const marginRight = viewportParams.margins?.marginRight ?? 0;
  const marginBottom = viewportParams.margins?.marginBottom ?? 0;
  const marginLeft = viewportParams.margins?.marginLeft ?? 0;

  const layerW = viewportParams.width - marginLeft - marginRight;
  const layerH = viewportParams.height - marginTop - marginBottom;
  const layerAspectRatio = layerW / layerH;

  let xScale = 1.0;
  let yScale = 1.0;
  if (viewportParams.aspectRatioMode === "Contain") {
    if (layerAspectRatio > 1.0) xScale = layerAspectRatio;
    else if (layerAspectRatio < 1.0) yScale = 1.0 / layerAspectRatio;
  } else if (viewportParams.aspectRatioMode === "Cover") {
    if (layerAspectRatio > 1.0) yScale = 1.0 / layerAspectRatio;
    else if (layerAspectRatio < 1.0) xScale = layerAspectRatio;
  }

  let xAlignTranslation = 0.0;
  let yAlignTranslation = 0.0;
  if (viewportParams.aspectRatioAlignmentMode === "Start") {
    xAlignTranslation = xScale - 1.0;
    yAlignTranslation = yScale - 1.0;
  } else if (viewportParams.aspectRatioAlignmentMode === "End") {
    xAlignTranslation = 1.0 - xScale;
    yAlignTranslation = 1.0 - yScale;
  }

  const xAdj = xScale - 1.0;
  const yAdj = yScale - 1.0;

  const xMin = ((-translateX - 1.0 - xAdj + xAlignTranslation) / zoomX + 1.0) / 2.0;
  const xMax = ((-translateX + 1.0 + xAdj + xAlignTranslation) / zoomX + 1.0) / 2.0;
  const yMin = ((-translateY - 1.0 - yAdj + yAlignTranslation) / zoomY + 1.0) / 2.0;
  const yMax = ((-translateY + 1.0 + yAdj + yAlignTranslation) / zoomY + 1.0) / 2.0;

  return { xMin, xMax, yMin, yMax };
}

/**
 * Given data bounds, compute the corresponding camera matrix.
 * Missing bound values are filled in from `prevCameraMatrix` so partial updates
 * (e.g. panning only the X axis) work without resetting the other axis.
 *
 * Bounds are expressed in normalized data coordinates where 0.0 is the
 * left/top edge of the data and 1.0 is the right/bottom edge.
 *
 * @param bounds - The desired visible data range. Any omitted fields are
 *   preserved from `prevCameraMatrix`.
 * @param prevCameraMatrix - The current camera matrix, used to fill in any
 *   omitted bound values. Typically the matrix returned by a previous call or
 *   the renderer's current state.
 * @param viewportParams - Describes the canvas size, aspect-ratio handling, and
 *   margins so the zoom level can be computed correctly.
 * @returns A column-major 4×4 `Float32Array` suitable for passing directly to
 *   the renderer as the camera matrix.
 *
 * @example
 * ```ts
 * import { getCameraMatrixFromBounds } from "@pluot/core";
 *
 * const viewport = {
 *   width: 800,
 *   height: 600,
 *   aspectRatioMode: "Contain",
 *   aspectRatioAlignmentMode: "Center",
 *   margins: { marginTop: 10, marginRight: 10, marginBottom: 10, marginLeft: 10 },
 * };
 *
 * // Identity camera matrix (full data range visible).
 * const identity = new Float32Array([
 *   1, 0, 0, 0,
 *   0, 1, 0, 0,
 *   0, 0, 1, 0,
 *   0, 0, 0, 1,
 * ]);
 *
 * // Zoom into the top-left quadrant of the data.
 * const cameraMatrix = getCameraMatrixFromBounds(
 *   { xMin: 0.0, xMax: 0.5, yMin: 0.0, yMax: 0.5 },
 *   identity,
 *   viewport,
 * );
 *
 * // Pan to shift only the X axis, keeping Y unchanged.
 * const pannedMatrix = getCameraMatrixFromBounds(
 *   { xMin: 0.1, xMax: 0.6 },
 *   cameraMatrix,
 *   viewport,
 * );
 * ```
 */
export function getCameraMatrixFromBounds(bounds: Bounds, prevCameraMatrix: Float32Array, viewportParams: ViewportParams): Float32Array {
  // Fill in missing bounds from the previous camera matrix.
  const currentBounds = getBounds(prevCameraMatrix, viewportParams);
  const xMin = bounds.xMin ?? currentBounds.xMin;
  const xMax = bounds.xMax ?? currentBounds.xMax;
  const yMin = bounds.yMin ?? currentBounds.yMin;
  const yMax = bounds.yMax ?? currentBounds.yMax;

  const marginTop = viewportParams.margins?.marginTop ?? 0;
  const marginRight = viewportParams.margins?.marginRight ?? 0;
  const marginBottom = viewportParams.margins?.marginBottom ?? 0;
  const marginLeft = viewportParams.margins?.marginLeft ?? 0;

  const layerW = viewportParams.width - marginLeft - marginRight;
  const layerH = viewportParams.height - marginTop - marginBottom;
  const layerAspectRatio = layerW / layerH;

  let xScale = 1.0;
  let yScale = 1.0;
  if (viewportParams.aspectRatioMode === "Contain") {
    if (layerAspectRatio > 1.0) xScale = layerAspectRatio;
    else if (layerAspectRatio < 1.0) yScale = 1.0 / layerAspectRatio;
  } else if (viewportParams.aspectRatioMode === "Cover") {
    if (layerAspectRatio > 1.0) yScale = 1.0 / layerAspectRatio;
    else if (layerAspectRatio < 1.0) xScale = layerAspectRatio;
  }

  let xAlignTranslation = 0.0;
  let yAlignTranslation = 0.0;
  if (viewportParams.aspectRatioAlignmentMode === "Start") {
    xAlignTranslation = xScale - 1.0;
    yAlignTranslation = yScale - 1.0;
  } else if (viewportParams.aspectRatioAlignmentMode === "End") {
    xAlignTranslation = 1.0 - xScale;
    yAlignTranslation = 1.0 - yScale;
  }

  const xAdj = xScale - 1.0;
  const yAdj = yScale - 1.0;

  const xRange = xMax - xMin;
  const yRange = yMax - yMin;

  let zoomX = (1.0 + xAdj) / xRange;
  let zoomY = (1.0 + yAdj) / yRange;

  // When aspect ratio is ignored, zoom each axis independently.
  // Otherwise take the minimum so all requested data fits within the viewport.
  if (viewportParams.aspectRatioMode !== "Ignore") {
    zoomX = zoomY = Math.min(zoomX, zoomY);
  }

  // Invert the getBounds translation equations:
  //   min + max = (-translate + align) / zoom + 1.0
  // So: translate = align - zoom * ((min + max) - 1.0)
  const translateX = xAlignTranslation - zoomX * ((xMin + xMax) - 1.0);
  const translateY = yAlignTranslation - zoomY * ((yMin + yMax) - 1.0);

  return new Float32Array([
    zoomX, 0.0,   0.0, 0.0,
    0.0,   zoomY, 0.0, 0.0,
    0.0,   0.0,   1.0, 0.0,
    translateX, translateY, 0.0, 1.0,
  ]);
}

File: bindings-python/pluot_core/src/pluot_core/viewport.py
from __future__ import annotations
from typing import Literal, Optional
from dataclasses import dataclass, field
import numpy as np

# TODO: auto-generate these types from the Rust side: https://github.com/keller-mark/pluot/issues/133
AspectRatioMode = Literal["Ignore", "Contain", "Cover"]
AspectRatioAlignmentMode = Literal["Center", "Start", "End"]


@dataclass
class Margins:
    margin_top: float = 0.0
    margin_right: float = 0.0
    margin_bottom: float = 0.0
    margin_left: float = 0.0


@dataclass
class ViewportParams:
    width: float
    height: float
    aspect_ratio_mode: AspectRatioMode
    aspect_ratio_alignment_mode: AspectRatioAlignmentMode
    margins: Optional[Margins] = field(default=None)


@dataclass
class Bounds:
    # Each value is optional.
    # When an entire dimension is omitted (X or Y),
    # use the current camera settings for that dimension.
    # When a single value is omitted (e.g., x_min), ensure the resulting camera matrix keeps this boundary unchanged.
    x_min: Optional[float] = None
    x_max: Optional[float] = None
    y_min: Optional[float] = None
    y_max: Optional[float] = None


def _get_scales_and_align_translations(viewport_params: ViewportParams):
    margins = viewport_params.margins
    margin_top = margins.margin_top if margins else 0.0
    margin_right = margins.margin_right if margins else 0.0
    margin_bottom = margins.margin_bottom if margins else 0.0
    margin_left = margins.margin_left if margins else 0.0

    layer_w = viewport_params.width - margin_left - margin_right
    layer_h = viewport_params.height - margin_top - margin_bottom
    layer_aspect_ratio = layer_w / layer_h

    x_scale = 1.0
    y_scale = 1.0
    if viewport_params.aspect_ratio_mode == "Contain":
        if layer_aspect_ratio > 1.0:
            x_scale = layer_aspect_ratio
        elif layer_aspect_ratio < 1.0:
            y_scale = 1.0 / layer_aspect_ratio
    elif viewport_params.aspect_ratio_mode == "Cover":
        if layer_aspect_ratio > 1.0:
            y_scale = 1.0 / layer_aspect_ratio
        elif layer_aspect_ratio < 1.0:
            x_scale = layer_aspect_ratio

    x_align_translation = 0.0
    y_align_translation = 0.0
    if viewport_params.aspect_ratio_alignment_mode == "Start":
        x_align_translation = x_scale - 1.0
        y_align_translation = y_scale - 1.0
    elif viewport_params.aspect_ratio_alignment_mode == "End":
        x_align_translation = 1.0 - x_scale
        y_align_translation = 1.0 - y_scale

    return x_scale, y_scale, x_align_translation, y_align_translation


def get_bounds(camera_matrix: np.ndarray, viewport_params: ViewportParams) -> Bounds:
    """Calculate the visible data range based on camera view and viewport parameters."""
    zoom_x = camera_matrix[0]
    zoom_y = camera_matrix[5]
    translate_x = camera_matrix[12]
    translate_y = camera_matrix[13]

    x_scale, y_scale, x_align_translation, y_align_translation = _get_scales_and_align_translations(viewport_params)

    x_adj = x_scale - 1.0
    y_adj = y_scale - 1.0

    x_min = ((-translate_x - 1.0 - x_adj + x_align_translation) / zoom_x + 1.0) / 2.0
    x_max = ((-translate_x + 1.0 + x_adj + x_align_translation) / zoom_x + 1.0) / 2.0
    y_min = ((-translate_y - 1.0 - y_adj + y_align_translation) / zoom_y + 1.0) / 2.0
    y_max = ((-translate_y + 1.0 + y_adj + y_align_translation) / zoom_y + 1.0) / 2.0

    return Bounds(x_min=x_min, x_max=x_max, y_min=y_min, y_max=y_max)


def get_camera_matrix_from_bounds(bounds: Bounds, prev_camera_matrix: np.ndarray, viewport_params: ViewportParams) -> np.ndarray:
    """Given data bounds, compute the corresponding camera matrix.
    Missing bound values are filled in from prev_camera_matrix.
    """
    current_bounds = get_bounds(prev_camera_matrix, viewport_params)
    x_min = bounds.x_min if bounds.x_min is not None else current_bounds.x_min
    x_max = bounds.x_max if bounds.x_max is not None else current_bounds.x_max
    y_min = bounds.y_min if bounds.y_min is not None else current_bounds.y_min
    y_max = bounds.y_max if bounds.y_max is not None else current_bounds.y_max

    x_scale, y_scale, x_align_translation, y_align_translation = _get_scales_and_align_translations(viewport_params)

    x_adj = x_scale - 1.0
    y_adj = y_scale - 1.0

    x_range = x_max - x_min
    y_range = y_max - y_min

    zoom_x = (1.0 + x_adj) / x_range
    zoom_y = (1.0 + y_adj) / y_range

    # When aspect ratio is ignored, zoom each axis independently.
    # Otherwise take the minimum so all requested data fits within the viewport.
    if viewport_params.aspect_ratio_mode != "Ignore":
        zoom_x = zoom_y = min(zoom_x, zoom_y)

    # Invert the get_bounds translation equations:
    #   min + max = (-translate + align) / zoom + 1.0
    # So: translate = align - zoom * ((min + max) - 1.0)
    translate_x = x_align_translation - zoom_x * ((x_min + x_max) - 1.0)
    translate_y = y_align_translation - zoom_y * ((y_min + y_max) - 1.0)

    return np.array([
        zoom_x, 0.0,   0.0, 0.0,
        0.0,   zoom_y, 0.0, 0.0,
        0.0,   0.0,    1.0, 0.0,
        translate_x, translate_y, 0.0, 1.0,
    ], dtype=np.float32)
